### ISSUE TITLE
Misuses of cryptographic APIs

### ISSUE THREAD

**Issue by mhp0rtal** [jwtk/jjwt#459]
Hi
The following lines have cryptographic API misuses. <br> File name => api/src/main/java/io/jsonwebtoken/SigningKeyResolverAdapter.java: Line number => 51: API name => SecretKeySpec:<BR> File name => api/src/main/java/io/jsonwebtoken/security/Keys.java: Line number => 70: API name => SecretKeySpec:<BR> File name => api/src/main/java/io/jsonwebtoken/SigningKeyResolverAdapter.java: Line number => 62: API name => SecretKeySpec:<BR> File name => api/src/main/java/io/jsonwebtoken/SigningKeyResolverAdapter.java: Line number => 62: API name => SecretKeySpec:<BR> File name => api/src/main/java/io/jsonwebtoken/security/Keys.java: Line number => 70: API name => SecretKeySpec:<BR> File name => api/src/main/java/io/jsonwebtoken/SigningKeyResolverAdapter.java: Line number => 51: API name => SecretKeySpec:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/RsaProvider.java: Line number => 186: API name => KeyPairGenerator:First parameter (with value 4096) should be any of {1024, 2048, 4096}<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java: Line number => 134: API name => SecretKeySpec:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java: Line number => 353: API name => SecretKeySpec:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/SignatureProvider.java: Line number => 73: API name => Signature:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/RsaSignatureValidator.java: Line number => 63: API name => Signature:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java: Line number => 353: API name => SecretKeySpec:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveSignatureValidator.java: Line number => 63: API name => Signature:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/MacSigner.java: Line number => 64: API name => Mac:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/DefaultJwtBuilder.java: Line number => 134: API name => SecretKeySpec:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/MacSigner.java: Line number => 32: API name => SecretKeySpec:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/MacSigner.java: Line number => 32: API name => SecretKeySpec:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveProvider.java: Line number => 134: API name => KeyPairGenerator:First parameter (with value "EC") should be any of {RSA, DSA, Diffie-Hellman}<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/RsaProvider.java: Line number => 193: API name => KeyPairGenerator:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveSigner.java: Line number => 57: API name => Signature:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/RsaSigner.java: Line number => 56: API name => Signature:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveProvider.java: Line number => 136: API name => KeyPairGenerator:First parameter (with value "EC") should be any of {RSA, DSA, Diffie-Hellman}<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/EllipticCurveProvider.java: Line number => 142: API name => KeyPairGenerator:<BR> File name => impl/src/main/java/io/jsonwebtoken/impl/crypto/MacProvider.java: Line number => 92: API name => KeyGenerator:<BR>

**Comment by lhazlewood** [comment]
What are the misuses?  Where did you get this information?  What is your justification?

Just pasting a list of random information isn't too helpful.  I'm likely to close this without any further information.

**Comment by DanFTRX** [comment]
@lhazlewood This guy(bot?) posted 900 issues to 900 repo's all listing "cryptographic API misuses"  It looks like some sort of bot that ran an overzealous static analysis without any context of use.

**Comment by lhazlewood** [comment]
@mhp0rtal do you have any context here?  How do you know they're misuses? What is the justification for each of these line items from your static analysis tool?

**Comment by lhazlewood** [comment]
@DanFTRX thanks for the heads up - let's see if we get any relevant information.

**Comment by mhp0rtal** [comment]
Hi,
Thanks for your attention.
I reported the cryptographic misuses due to my recent research on how developers employ Java Cryptography Architecture (JCA) APIs in Java applications. I utilized some static analysis tools intended for checking cryptographic practices. The aforementioned misuses in those files were obtained from the results of the tools. The severity of the bug is relative. For instance, Even though MD5 is listed on Java documentation, MD5 is not recommended to use in any language anymore as it is considered insecure. An attacker can try billions of candidate passwords per second on a single GPU. Nevertheless, depending on the scenario the same hash algorithm may not be a problem for file integrity purpose.
There are several instances of cryptographic misuses which could lead to a security vulnerability:
*Do not use ECB mode for encryption.
*Do not use a non-random IV for CBC encryption
*Do not use constant encryption keys.
*Do not use constant salts for PBE.
*Do not use fewer than 1,000 iterations for PBE.
*Do not use static seeds to seed SecureRandom(·)
It seems most of your misuses are related to SecretKeySpec, I suggest you run CryptoLint or the CogniCrypt tool on your code to get for some suggestions.
